File: src/Systems/Zig/example3.zig
const std = @import("std");

const Report = struct {
    generate: fn () void,
};

fn generatePDFReport() void {
    std.debug.print("Generating PDF report\n", .{});
}

fn generateHTMLReport() void {
    std.debug.print("Generating HTML report\n", .{});
}

const PDFReportFactory = struct {
    pub fn createReport() Report {
        return Report{
            .generate = generatePDFReport,
        };
    }
};

const HTMLReportFactory = struct {
    pub fn createReport() Report {
        return Report{
            .generate = generateHTMLReport,
        };
    }
};

fn useFactory(factory: anytype) void {
    const report = factory.createReport();
    report.generate();
}

pub fn main() void {
    useFactory(PDFReportFactory);
    useFactory(HTMLReportFactory);
}

File: src/Niche/Nim/example3.nim
type
  Report = ref object of RootObj
  proc generate(self: Report)

type
  PDFReport = ref object of Report
  HTMLReport = ref object of Report

proc generate(self: PDFReport) =
  echo "Generating PDF report"

proc generate(self: HTMLReport) =
  echo "Generating HTML report"

type
  ReportFactory = ref object of RootObj
  proc createReport(self: ReportFactory): Report

type
  PDFReportFactory = ref object of ReportFactory
  HTMLReportFactory = ref object of ReportFactory

proc createReport(self: PDFReportFactory): Report =
  PDFReport()

proc createReport(self: HTMLReportFactory): Report =
  HTMLReport()

proc useFactory(factory: ReportFactory) =
  let report = factory.createReport()
  report.generate()

# Usage
let pdfFactory = PDFReportFactory()
let htmlFactory = HTMLReportFactory()
useFactory(pdfFactory)
useFactory(htmlFactory)
